### PR TITLE
fix: check for truthy value of deleteness & act accordingly

### LINKED ISSUE #474: "These linked pull requests are closed" should only be commented if any are counted
"These linked pull requests are closed" should only be commented if any are counted. I've seen in Draeieg's quality assurance examples it posting a comment like that without closing any. 

https://github.com/ubiquity/ubiquibot/blob/48e006da17c46cc172263fe9a3924cd90a0f3bcb/src/handlers/assign/action.ts#L78-L83

### PR DESCRIPTION
**Pull request by b4s36t4**
Resolves #474

### CODE REVIEW

File: src/helpers/issue.ts [hunk 1]
```diff
@@ -281,14 +281,15 @@ export const closePullRequest = async (pull_number: number) => {
   const payload = context.payload as Payload;
   const logger = getLogger();
   try {
-    await getBotContext().octokit.rest.pulls.update({
+    return await getBotContext().octokit.rest.pulls.update({
```
Comment (pavlovcik) [inline src/helpers/issue.ts:284]: It is unclear to me what the return value is of this function from this code review view.

File: src/handlers/assign/action.ts [hunk 2]
```diff
@@ -75,10 +75,23 @@ export const closePullRequestForAnIssue = async (): Promise<void> => {
...
   let comment = `These linked pull requests are closed: `;
+  let noPullRequestsComment = `Failed To delete these pull requests: `;
+
+  let deleteCount = 0;
   for (let i = 0; i < prs.length; i++) {
-    await closePullRequest(prs[i].number);
+    const deleted = await closePullRequest(prs[i].number);
+    if (!deleted) {
+      noPullRequestsComment += ` <a href="${prs[i]._links.html.href}">#${prs[i].number}</a> `;
+      deleteCount++;
```
Comment (pavlovcik) [inline src/handlers/assign/action.ts:87]: It seems unusual to me that you're saying "if nothing was deleted, add to the delete count" is that sound logic?

Comment (b4s36t4) [reply src/handlers/assign/action.ts:87]: haha, yes. poor naming, will update.

File: src/helpers/issue.ts (hunk 1 above)
Comment (b4s36t4) [reply src/helpers/issue.ts]: I have updated code, please check.

**Comment by b4s36t4** [comment]
@pavlovcik Updated the PR, review once you can.

**Comment by b4s36t4** [comment]
By logs I feel this issue might be related to some bot configuration, like permission or something.
Screenshot 2023-07-07 at 3 20 50 AM

**Comment by pavlovcik** [comment]
@0xcodercrane

**Comment by 0xcodercrane** [comment]
> By logs I feel this issue might be related to some bot configuration, like permission or something. Screenshot 2023-07-07 at 3 20 50 AM

would you please give us more details? A couple of reasons for that issue but hard to say exactly what the root cause is from the screenshot.

here are the possible reasons for that.
1/ Limited permission of your GitHub app.
2/ Probot framework issue --- A few octokit APIs doesn't get supported by Probot due to its internal issue. you can take a look at the [probot issue](https://github.com/probot/probot/issues/1243). It happened to us as well in the past.
In this case, we should figure out which method call throws the above error and replace it with another one.

File: src/handlers/assign/action.ts [hunk 3]
```diff
@@ -75,10 +75,24 @@ export const closePullRequestForAnIssue = async (): Promise<void> => {
   const prs = await getOpenedPullRequestsForAnIssue(payload.issue.number, "");
   if (!prs.length) return;
   logger.info(`Opened prs for this issue: ${JSON.stringify(prs)}`);
+
   let comment = `These linked pull requests are closed: `;
+  let noPullRequestsComment = `Failed To close these pull requests: `;
+
+  let deleteCount = 0;
   for (let i = 0; i < prs.length; i++) {
```
Comment (0xcodercrane) [inline src/handlers/assign/action.ts:83]:
```suggestion
  for (const pr of prs) {
```

File: src/handlers/assign/action.ts [hunk 4]
```diff
@@ -75,10 +75,24 @@ export const closePullRequestForAnIssue = async (): Promise<void> => {
...
-    await closePullRequest(prs[i].number);
+    const deleted = await closePullRequest(prs[i].number);
```
Comment (0xcodercrane) [inline src/handlers/assign/action.ts:84]:
```suggestion
    const deleted = await closePullRequest(pr.number);
```

File: src/handlers/assign/action.ts [hunk 5]
```diff
@@ -75,10 +75,24 @@ export const closePullRequestForAnIssue = async (): Promise<void> => {
...
+    if (!deleted) {
+      noPullRequestsComment += ` <a href="${prs[i]._links.html.href}">#${prs[i].number}</a> `;
```
Comment (0xcodercrane) [inline src/handlers/assign/action.ts:86]:
```suggestion
      noPullRequestsComment += ` <a href="${pr._links.html.href}">#${pr.number}</a> `;
```

File: src/handlers/assign/action.ts [hunk 6]
```diff
@@ -75,10 +75,24 @@ export const closePullRequestForAnIssue = async (): Promise<void> => {
...
+      continue;
+    }
     comment += ` <a href="${prs[i]._links.html.href}">#${prs[i].number}</a> `;
```
Comment (0xcodercrane) [inline src/handlers/assign/action.ts:89]:
```suggestion
    comment += ` <a href="${pr._links.html.href}">#${pr.number}</a> `;
```

**Review by 0xcodercrane — CHANGES_REQUESTED** [review]
Some of the suggestions should've happened in previous PR reviews, not in your PR. Unfortunately, I didn't catch them up but want you to accept the suggestions if its not a big deal to you @b4s36t4

File: src/handlers/assign/action.ts [hunk 7]
```diff
@@ -75,10 +75,24 @@ export const closePullRequestForAnIssue = async (): Promise<void> => {
...
+    deleteCount++;
+  }
+  if (deleteCount === 0) {
+    // If no PR is closed
+    await addCommentToIssue(noPullRequestsComment, payload.issue.number);
+    return;
```
Comment (wannacfuture) [inline src/handlers/assign/action.ts:95]: Also, not needed.

File: src/handlers/assign/action.ts [hunk 8]
```diff
@@ -75,10 +75,24 @@ export const closePullRequestForAnIssue = async (): Promise<void> => {
   const prs = await getOpenedPullRequestsForAnIssue(payload.issue.number, "");
   if (!prs.length) return;
   logger.info(`Opened prs for this issue: ${JSON.stringify(prs)}`);
+
   let comment = `These linked pull requests are closed: `;
+  let noPullRequestsComment = `Failed To close these pull requests: `;
+
+  let deleteCount = 0;
```
Comment (wannacfuture) [inline src/handlers/assign/action.ts:82]: Seems not needed.
In the above code, if(!pr.length) return;
So it won't be reached here.

**Review by wannacfuture — CHANGES_REQUESTED** [review]

**Comment by b4s36t4** [comment]
@wannacfuture it's not about handling the PR's count it's about handling the PR delete count. There might be some cases, like if one of the member from org removes `Write` Permission to pull request. Even if the PR wasn't closed bot will send as Completed. Isn't that an issue @pavlovcik ?

**Comment by b4s36t4** [comment]
And I don't this was added, we should let the issue members know the status of attached PRs.

**Comment by pavlovcik** [comment]
I don't understand your point

**Comment by b4s36t4** [comment]
@pavlovcik I think @wannacfuture is mentioning that if no open PRs are linked to issue bot won't message in the issue stating closing the PRs. But what if the Closing PR's failed? Maybe permission error as I have shown above? which in this case is messaging PR's were closed but actually it isn't closed

**Comment by wannacfuture** [comment]
> which in this case is messaging PR's were closed but actually it isn't closed

😊Thanks. But the mentioned QA is the example QA for old one(not including `if(!pr.length) return;`), not updated PR.
The updated one works for sure. @b4s36t4

**Comment by b4s36t4** [comment]
@pavlovcik @wannacfuture @0xcodercrane Updated the PR please review it again. This time I have added the check to send both failed and success PR's message. 

Let's say there are 5 PRs out of 2 were closed successfully, bot now will send 2 messages one with failed PR's one with success PR's. Let me know if any changes are required.

**Pull request closed without merging**